kbmadm/kbmadm.c

```diff
@@ -0,0 +1,1123 @@
+	boolean_t opt_p = B_FALSE;
+	boolean_t opt_v = B_FALSE;
+
+	while ((c = getopt(argc, argv, "pv")) != -1) {
+		switch (c) {
+		case 'p':
+			opt_p = B_TRUE;
+			break;
+		case 'v':
+			opt_v = B_TRUE;
+			break;
+		default:
+			(void) fprintf(stderr, "Invalid flag -%c\n", optopt);
+			usage();
+		}
+	}
+
+	if ((ret = req_new(KBM_CMD_LIST_RECOVERY, &req)) != ERRF_OK)
+		return (ret);
+
+	if ((ret = send_request(req, &resp)) != ERRF_OK)
+		goto done;
+
+	if ((ret = envlist_lookup_nvlist(resp, KBM_NV_CONFIGS,
+	    &cfgs)) != ERRF_OK) {
+		goto done;
+	}
+
+	for (pair = nvlist_next_nvpair(cfgs, NULL); pair != NULL;
+	    pair = nvlist_next_nvpair(cfgs, pair)) {
+		const char *name = nvpair_name(pair);
+		nvlist_t *cfg = NULL;
+		nvlist_t **cfgs = NULL;
+		uint_t ncfgs = 0;
+		uint8_t *hash = NULL;
+		uint_t hashlen = 0;
+
+		if (nvpair_type(pair) != DATA_TYPE_NVLIST) {
+			ret = errf("InternalError", NULL,
+			    "kbmd response was not an nvlist");
+			goto done;
+		}
+
+		VERIFY0(nvpair_value_nvlist(pair, &cfg));
+
+		if ((ret = envlist_lookup_uint8_array(cfg, KBM_NV_CONFIG_HASH,
+		    &hash, &hashlen)) != ERRF_OK)
+			goto done;
+
+		char hashstr[hashlen * 2 + 1];
+		tohex(hash, (size_t)hashlen, hashstr, hashlen * 2 + 1);
+
+		if (opt_p) {
+			(void) printf("%s:%s\n", name, hashstr);
+			continue;
+		}
+
+		if ((ret = envlist_lookup_nvlist_array(cfg, KBM_NV_CONFIGS,
+		    &cfgs, &ncfgs)) != ERRF_OK)
+			goto done;
+
+		(void) printf("%s config (%s):\n", name, hashstr);
+
+		if ((ret = show_configs(cfgs, ncfgs, opt_v)) != ERRF_OK)
+			goto done;
+	}
+
+done:
+	nvlist_free(req);
+	nvlist_free(resp);
+	return (ret);
+}
+
+static errf_t *
+do_activate_recovery(int argc, char **argv)
+{
+	errf_t *ret = ERRF_OK;
+	const char *dataset = NULL;
+	nvlist_t *req = NULL;
+	nvlist_t *resp = NULL;
+
+	if ((dataset = argv[1]) == NULL) {
+		return (errf("ArgumentError", NULL, "dataset name is missing"));
+	}
+
+	if ((ret = req_new(KBM_CMD_ACTIVATE_RECOVERY, &req)) != ERRF_OK)
+		goto done;
+
+	if (dataset != NULL &&
+	    (ret = envlist_add_string(req, KBM_NV_DATASET,
+	    dataset)) != ERRF_OK)
+		goto done;
+
+	ret = send_request(req, &resp);
+
+done:
+	nvlist_free(req);
+	nvlist_free(resp);
+	return (ret);
+}
+
+static errf_t *
+do_cancel_recovery(int argc, char **argv)
+{
+	errf_t *ret = ERRF_OK;
+	const char *dataset = NULL;
+	nvlist_t *req = NULL;
+	nvlist_t *resp = NULL;
+
+	if ((dataset = argv[optind]) == NULL) {
+		return (errf("ArgumentError", NULL, "dataset name is missing"));
+	}
+
+	if ((ret = req_new(KBM_CMD_CANCEL_RECOVERY, &req)) != ERRF_OK)
+		goto done;
+
+	if (dataset != NULL &&
+	    (ret = envlist_add_string(req, KBM_NV_DATASET,
+	    dataset)) != ERRF_OK)
+		goto done;
+
+	ret = send_request(req, &resp);
+
+done:
+	nvlist_free(req);
+	nvlist_free(resp);
+	return (ret);
+}
+
+static errf_t *
+do_set_syspool(int argc, char **argv)
+{
+	errf_t *ret = ERRF_OK;
+	nvlist_t *req = NULL;
+	nvlist_t *resp = NULL;
+
+	if (argc < 2) {
+		usage();
```

Comment:
I'll add a comment that it's intentionally undocumented (the systoken and sys pool were for some later stuff Alex wanted, and really today just act as default values, so it might not be worth keeping them anyway).

Reply:
So maybe better to delete if there's no consumer so that if we decide to go a different direction in the future there is no conversion required.

Reply:
I'll keep the sys pool around mostly for diagnostic purposes, but also use it to set the GUID for the system token (which we use as a default value, but only store the GUID instead of anything else -- that gets rid of most of the extra logic).

Reply:
confirmed